src/prefect/_internal/concurrency/event_loop.py

```diff
@@ -33,6 +33,10 @@ def call_in_loop(
 ) -> T:
     """
     Run a synchronous call in event loop's thread from another thread.
+
+    This function is blocking and not safe to call from an asynchronous context.
+
+    Returns the result of the call.
     """
     future = call_soon_in_loop(__loop, __fn, *args, **kwargs)
```

Comment:
So, is the change here that instead of `call_soon_in_loop` always calling `call_soon_threadsafe`, `call_soon_in_loop` determines whether we're in the main thread and if not then would call `call_soon_threadsafe` to schedule the callback?

Reply:
This is a bit of a minor change I snuck in here, it doesn't actually change any behavior relevant for this pull request (shame on me). There's basically `loop.call_soon` or `loop.call_soon_threadsafe` — one's for using from the event loop thread and the other let's you schedule things from other threads. By checking if the current event loop is the one you're scheduling something on, we can switch to the threadsafe implementation when needed and use the normal one by default — this avoids whatever synchronization overhead they use to make the call threadsafe when we don't need it.

Reply:
I'm talking about the change in https://github.com/PrefectHQ/prefect/pull/7897/files#diff-1d13a991aaa8d66f948979d55a7e7dad20ee565c9a94a88e0c80aaa25cfd741cR71-R76 btw.

The fix for the deadlock in this PR is calling `call_soon_in_loop` at https://github.com/PrefectHQ/prefect/pull/7897/files#diff-fb1d7293ffe4010ef6082fa1d4c5235a52a8129e6e33542c808f2edde3332ce7R53 instead of calling `call_in_loop` — which blocks until completion and can fail when there are dependent waiters.

Reply:
> By checking if the current event loop is the one you're scheduling something on, we can switch to the threadsafe implementation when needed but use the unsafe one otherwise — this avoids whatever synchronization overhead they use to make the call threadsafe when we don't need it.

Got it, that's about what I thought the reasoning was 🧐